edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/device/DeviceMessageHandler.cs

```diff
@@ -179,9 +181,24 @@ public async Task SendGetTwinRequest(string correlationId)
             }
         }
 
-        public Task ProcessDeviceMessageAsync(IMessage message) => this.edgeHub.ProcessDeviceMessage(this.Identity, message);
+        public Task ProcessDeviceMessageAsync(IMessage message)
+        {
+            this.modelId.ForEach(modelId => message.SystemProperties[SystemProperties.ModelId] = modelId);
+            return this.edgeHub.ProcessDeviceMessage(this.Identity, message);
+        }
+
+        public Task ProcessDeviceMessageBatchAsync(IEnumerable<IMessage> messages)
+        {
+            this.modelId.ForEach(modelId =>
+            {
+                foreach (IMessage message in messages)
+                {
+                    message.SystemProperties[SystemProperties.ModelId] = modelId;
```

Comment:
>message.SystemProperties[SystemProperties.ModelId] = modelId; [](start = 20, length = 61)

I think for edge-specific metadata, we avoid putting it as an actual message header.  Instead it gets wrapped into the various IMessage implementations.

The danger of putting the ID as a header is that we're modifying the original message, and it'll be difficult to maintain on the sending end the list of headers that we need to strip back out.

Reply:
So you believe we should add it as a field in the IMessage implementations and just deal with whatever conversions need to be done as it flows through routing?

I was planning on adding logic on the sending side that would just strip out any SystemProperty with SystemProperties.ModelID in it before sending. What difficulty do you foresee?

Reply:
IMessage doesn't let you wrap another IMessage. You would do that only to make the SystemProperties immutable. But I think this is fine for now. This is an existing pattern. 

Also the System properties have to be explicitly set when sending the messages out, and this is not a property that can be set by the client.